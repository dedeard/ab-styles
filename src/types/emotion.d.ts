import '@emotion/react';
import { colors } from '../config';

export type Breakpoints = 'xs' | 'sm' | 'md' | 'lg' | 'xl';

/**
 * Flexbox properties
 */
export type Align =
  | 'baseline'
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'stretch';

export type Justify =
  | 'center'
  | 'flex-end'
  | 'flex-start'
  | 'space-around'
  | 'space-between';

export interface AlignProps {
  xs?: Align;
  sm?: Align;
  md?: Align;
  lg?: Align;
  xl?: Align;
}

export interface JustifyProps {
  xs?: Justify;
  sm?: Justify;
  md?: Justify;
  lg?: Justify;
  xl?: Justify;
}

export interface DefaultTheme {
  grid?: {
    space: number;
    container: Record<Breakpoints, number | string>;
    columns: Record<Breakpoints, number>;
    breakpoints: Record<Breakpoints, number>;
  };
  colors?: typeof colors;
}
export interface StyleProps {
  theme: DefaultTheme;
}

declare module '@emotion/react' {
  export interface Theme extends DefaultTheme {}
}
