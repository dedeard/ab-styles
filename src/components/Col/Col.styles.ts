import { css } from '@emotion/react';

import { config } from '../../config';
import { responsive } from '../../utils';

import type { ColProps } from '.';
import type { DefaultTheme, StyleProps } from '../../types/emotion';

export const base = ({ theme }: StyleProps) => css`
  position: relative;
  flex: 1 0 0%;

  width: 100%;
  max-width: 100%;

  box-sizing: border-box;

  padding-left: ${config(theme).grid.space / 2}rem;
  padding-right: ${config(theme).grid.space / 2}rem;
`;

export const size = (props: ColProps & StyleProps) =>
  responsive(
    config(props.theme).grid.breakpoints,
    (breakpoint: keyof DefaultTheme['grid']['breakpoints']) =>
      props[breakpoint] &&
      `
        flex: 0 0 ${
          (props[breakpoint] / config(props.theme).grid.columns[breakpoint]) *
          100
        }%;
        max-width: ${
          (props[breakpoint] / config(props.theme).grid.columns[breakpoint]) *
          100
        }%;
    `
  );

export const offset = ({ theme, offset }: ColProps & StyleProps) =>
  offset &&
  responsive(
    config(theme).grid.breakpoints,
    (breakpoint: keyof DefaultTheme['grid']['breakpoints']) =>
      offset[breakpoint] &&
      `
        margin-left: ${
          offset[breakpoint] >= 0
            ? (offset[breakpoint] / config(theme).grid.columns[breakpoint]) *
              100
            : 0
        }%;
    `
  );
