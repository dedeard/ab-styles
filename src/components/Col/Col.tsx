import { CSSProperties, ReactNode } from 'react';

import styled from '@emotion/styled';
import { css } from '@emotion/react';

import { config, BREAKPOINTS } from '../../config';
import { media } from '../../utils';

import { Breakpoints, StyleProps } from '../../types/emotion';

export interface ColProps {
  xs?: number;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  offset?: Record<Breakpoints, number>;
  style?: CSSProperties;
  className?: string;
  children: ReactNode;
}

const baseStyle = ({ theme }: StyleProps) => css`
  label: col;

  position: relative;
  flex: 1 0 0%;

  width: 100%;
  max-width: 100%;

  box-sizing: border-box;

  ${BREAKPOINTS.map(
    (breakpoint) =>
      css`
        ${media(breakpoint)} {
          padding-left: ${config(theme).grid.gutter[breakpoint] / 2}rem;
          padding-right: ${config(theme).grid.gutter[breakpoint] / 2}rem;
        }
      `
  )}
`;

const sizeStyle = (props: ColProps & StyleProps) => {
  const { theme } = props;

  return css`
    ${BREAKPOINTS.map(
      (breakpoint) =>
        props[breakpoint] &&
        css`
          ${media(breakpoint)} {
            flex: 0 0
              ${(props[breakpoint] / config(theme).grid.columns[breakpoint]) *
              100}%;
            max-width: ${(props[breakpoint] /
              config(theme).grid.columns[breakpoint]) *
            100}%;
          }
        `
    )}
  `;
};

const offsetStyle = ({ theme, offset }: ColProps & StyleProps) =>
  offset &&
  typeof offset === 'object' &&
  BREAKPOINTS.filter((breakpoint) => offset[breakpoint]).map(
    (breakpoint) =>
      offset[breakpoint] &&
      css`
        ${media(breakpoint)} {
          margin-left: ${offset[breakpoint] >= 0
            ? (offset[breakpoint] / config(theme).grid.columns[breakpoint]) *
              100
            : 0}%;
        }
      `
  );

const Col = styled('div')<ColProps>(baseStyle, sizeStyle, offsetStyle);

export default Col;
