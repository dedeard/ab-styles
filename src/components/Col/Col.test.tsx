import React from 'react';
import { render } from '@testing-library/react';

import { Col } from './Col';
import { Breakpoints } from '../../types/types';

describe('Col', () => {
  it('should render with default styles', () => {
    const { container } = render(<Col>Col</Col>);

    expect(container.firstChild).toMatchSnapshot();
  });

  const sizes = [
    {
      xs: 1,
      sm: 2,
      md: 3,
      lg: 4,
      xl: 5,
    },
  ] as const;

  it.each(sizes)('should render with width styles for media %s', (size) => {
    const { container } = render(<Col {...size}>Col</Col>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it.each(sizes)('should render with offset styles', (size) => {
    const { container } = render(<Col offset={size}>Col</Col>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should handle 0 offset styles', () => {
    const offsets = { xs: 0, lg: 9 } as Record<Breakpoints, number>;
    const { container } = render(
      <Col md={8} offset={offsets}>
        Col
      </Col>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
