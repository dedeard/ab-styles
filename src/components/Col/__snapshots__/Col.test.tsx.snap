// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col should render with default styles 1`] = `
.grid-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

.emotion-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

<div
  class="EmotionGrid-Col emotion-0"
>
  Col
</div>
`;

exports[`Col should render with offset styles on breakpoint { xs: 1, sm: 2, md: 3, lg: 4, xl: 5 } 1`] = `
.grid-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 1em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 33.75em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 48em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 64em) {
  .grid-0 {
    -webkit-flex: 0 0 33.33333333333333%;
    -ms-flex: 0 0 33.33333333333333%;
    flex: 0 0 33.33333333333333%;
    max-width: 33.33333333333333%;
  }
}

@media (min-width: 71.25em) {
  .grid-0 {
    -webkit-flex: 0 0 41.66666666666667%;
    -ms-flex: 0 0 41.66666666666667%;
    flex: 0 0 41.66666666666667%;
    max-width: 41.66666666666667%;
  }
}

.emotion-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 1em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 33.75em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 48em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 64em) {
  .emotion-0 {
    -webkit-flex: 0 0 33.33333333333333%;
    -ms-flex: 0 0 33.33333333333333%;
    flex: 0 0 33.33333333333333%;
    max-width: 33.33333333333333%;
  }
}

@media (min-width: 71.25em) {
  .emotion-0 {
    -webkit-flex: 0 0 41.66666666666667%;
    -ms-flex: 0 0 41.66666666666667%;
    flex: 0 0 41.66666666666667%;
    max-width: 41.66666666666667%;
  }
}

<div
  class="EmotionGrid-Col emotion-0"
>
  Col
</div>
`;

exports[`Col should render with size styles on breakpoint { xs: 1, sm: 2, md: 3, lg: 4, xl: 5 } 1`] = `
.grid-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .grid-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 1em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 33.75em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 48em) {
  .grid-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 64em) {
  .grid-0 {
    -webkit-flex: 0 0 33.33333333333333%;
    -ms-flex: 0 0 33.33333333333333%;
    flex: 0 0 33.33333333333333%;
    max-width: 33.33333333333333%;
  }
}

@media (min-width: 71.25em) {
  .grid-0 {
    -webkit-flex: 0 0 41.66666666666667%;
    -ms-flex: 0 0 41.66666666666667%;
    flex: 0 0 41.66666666666667%;
    max-width: 41.66666666666667%;
  }
}

.emotion-0 {
  position: relative;
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
}

@media (min-width: 1em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 33.75em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 48em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 64em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 71.25em) {
  .emotion-0 {
    padding-left: 0.9375rem;
    padding-right: 0.9375rem;
  }
}

@media (min-width: 1em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 33.75em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 48em) {
  .emotion-0 {
    -webkit-flex: 0 0 25%;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }
}

@media (min-width: 64em) {
  .emotion-0 {
    -webkit-flex: 0 0 33.33333333333333%;
    -ms-flex: 0 0 33.33333333333333%;
    flex: 0 0 33.33333333333333%;
    max-width: 33.33333333333333%;
  }
}

@media (min-width: 71.25em) {
  .emotion-0 {
    -webkit-flex: 0 0 41.66666666666667%;
    -ms-flex: 0 0 41.66666666666667%;
    flex: 0 0 41.66666666666667%;
    max-width: 41.66666666666667%;
  }
}

<div
  class="EmotionGrid-Col emotion-0"
>
  Col
</div>
`;
